sdks/java/core/src/main/java/org/apache/beam/sdk/coders/JAXBCoder.java

```diff
@@ -68,9 +68,18 @@ private JAXBCoder(Class<T> jaxbClass) {
   public void encode(T value, OutputStream outStream, Context context)
       throws CoderException, IOException {
     try {
-      JAXBContext jaxbContext = getContext();
-      // TODO: Consider caching in a ThreadLocal if this impacts performance
-      Marshaller jaxbMarshaller = jaxbContext.createMarshaller();
+      ThreadLocal<Marshaller> jaxbMarshaller = new ThreadLocal<Marshaller>() {
```

Comment:
This needs to be stored in a field so it can live beyond the lifetime of the method; but it needs to be initialized if not present on the first call to encode, as we can't serialize the default Java `ThreadLocal`.

I would prefer to do this the same way AvroCoder does this, using a `ThreadLocal` that deserializes into an empty version (see `[EmptyOnDeserializationThreadLocal](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java#L182)`). Doing so would be best accomplished by refactoring the `EmptyOnDeserializationThreadLocal` into a class that can be reused within multiple coders.

You could alternatively use a `getMarshaller()`/`getUnmarshaller()` method which behaves similarly to `getContext()` (e.g. uses synchronization to set up